- **PuaMode** - Pua Enabled mode from the service
- **SMode** - This field is set to true when the device is known to be in 'S Mode', as in, Windows 10 S mode, where only Microsoft Store apps can be installed
- **IeVerIdentifier** - NA
- **SmartScreen** - This is the SmartScreen enabled string value from registry. This is obtained by checking in order, HKLM\SOFTWARE\Policies\Microsoft\Windows\System\SmartScreenEnabled and HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\SmartScreenEnabled. If the value exists but is blank, the value "ExistsNotSet" is sent in telemetry.
- **Firewall** - This attribute is true (1) for Windows 8.1 and above if windows firewall is enabled, as reported by the service.
- **UacLuaenable** - This attribute reports whether or not the "administrator in Admin Approval Mode" user type is disabled or enabled in UAC. The value reported is obtained by reading the regkey HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System\EnableLUA.
- **Census_MDC2FormFactor** - A grouping based on a combination of Device Census level hardware characteristics. The logic used to define Form Factor is rooted in business and industry standards and aligns with how people think about their device. (Examples: Smartphone, Small Tablet, All in One, Convertible...)
- **Census_DeviceFamily** - AKA DeviceClass. Indicates the type of device that an edition of the OS is intended for. Example values: Windows.Desktop, Windows.Mobile, and iOS.Phone
- **Census_OEMNameIdentifier** - NA
- **Census_OEMModelIdentifier** - NA
- **Census_ProcessorCoreCount** - Number of logical cores in the processor
- **Census_ProcessorManufacturerIdentifier** - NA
- **Census_ProcessorModelIdentifier** - NA
- **Census_ProcessorClass** - A classification of processors into high/medium/low. Initially used for Pricing Level SKU. No longer maintained and updated
- **Census_PrimaryDiskTotalCapacity** - Amount of disk space on primary disk of the machine in MB
- **Census_PrimaryDiskTypeName** - Friendly name of Primary Disk Type - HDD or SSD
- **Census_SystemVolumeTotalCapacity** - The size of the partition that the System volume is installed on in MB
- **Census_HasOpticalDiskDrive** - True indicates that the machine has an optical disk drive (CD/DVD)
- **Census_TotalPhysicalRAM** - Retrieves the physical RAM in MB
- **Census_ChassisTypeName** - Retrieves a numeric representation of what type of chassis the machine has. A value of 0 means xx
- **Census_InternalPrimaryDiagonalDisplaySizeInInches** - Retrieves the physical diagonal length in inches of the primary display
- **Census_InternalPrimaryDisplayResolutionHorizontal** - Retrieves the number of pixels in the horizontal direction of the internal display.
- **Census_InternalPrimaryDisplayResolutionVertical** - Retrieves the number of pixels in the vertical direction of the internal display
- **Census_PowerPlatformRoleName** - Indicates the OEM preferred power management profile. This value helps identify the basic form factor of the device
- **Census_InternalBatteryType** - NA
- **Census_InternalBatteryNumberOfCharges** - NA
- **Census_OSVersion** - Numeric OS version Example - 10.0.10130.0

In [32]:
import pandas as pd

ms = pd.read_csv('../../../sample_train.csv', low_memory=False)
nLinhas = ms.shape[0]

In [33]:
meio = ['PuaMode',
       'SMode', 'IeVerIdentifier', 'SmartScreen', 'Firewall', 'UacLuaenable',
       'Census_MDC2FormFactor', 'Census_DeviceFamily',
       'Census_OEMNameIdentifier', 'Census_OEMModelIdentifier',
       'Census_ProcessorCoreCount', 'Census_ProcessorManufacturerIdentifier',
       'Census_ProcessorModelIdentifier', 'Census_ProcessorClass',
       'Census_PrimaryDiskTotalCapacity', 'Census_PrimaryDiskTypeName',
       'Census_SystemVolumeTotalCapacity', 'Census_HasOpticalDiskDrive',
       'Census_TotalPhysicalRAM', 'Census_ChassisTypeName',
       'Census_InternalPrimaryDiagonalDisplaySizeInInches',
       'Census_InternalPrimaryDisplayResolutionHorizontal',
       'Census_InternalPrimaryDisplayResolutionVertical',
       'Census_PowerPlatformRoleName', 'Census_InternalBatteryType',
       'Census_InternalBatteryNumberOfCharges', 'Census_OSVersion']


In [34]:
ms = ms[meio]
ms.shape

(600000, 27)

In [35]:
### ver os valores nulos
print("Percentagem de NAs: ")
for col in ms:    
    percent_missing = ms[col].isnull().sum() * 100 / nLinhas
    if (percent_missing > 0):
        print(str(col) + ": " + str(percent_missing)+ "%")

Percentagem de NAs: 
PuaMode: 99.97716666666666%
SMode: 6.0155%
IeVerIdentifier: 0.6641666666666667%
SmartScreen: 35.611333333333334%
Firewall: 1.0311666666666666%
UacLuaenable: 0.1175%
Census_OEMNameIdentifier: 1.0676666666666668%
Census_OEMModelIdentifier: 1.1478333333333333%
Census_ProcessorCoreCount: 0.4821666666666667%
Census_ProcessorManufacturerIdentifier: 0.4821666666666667%
Census_ProcessorModelIdentifier: 0.4821666666666667%
Census_ProcessorClass: 99.58683333333333%
Census_PrimaryDiskTotalCapacity: 0.6208333333333333%
Census_PrimaryDiskTypeName: 0.14833333333333334%
Census_SystemVolumeTotalCapacity: 0.6206666666666667%
Census_TotalPhysicalRAM: 0.9355%
Census_ChassisTypeName: 0.0075%
Census_InternalPrimaryDiagonalDisplaySizeInInches: 0.5431666666666667%
Census_InternalPrimaryDisplayResolutionHorizontal: 0.5421666666666667%
Census_InternalPrimaryDisplayResolutionVertical: 0.5421666666666667%
Census_PowerPlatformRoleName: 0.0008333333333333334%
Census_InternalBatteryType: 71.052

In [36]:
### ver o tipo e a diversidade de valores
for col in ms:
    nome = ms[col].unique()
    print("A coluna " + str(col) + " tem: " + str(len(nome)) + " valores diferentes")  

A coluna PuaMode tem: 2 valores diferentes
A coluna SMode tem: 3 valores diferentes
A coluna IeVerIdentifier tem: 190 valores diferentes
A coluna SmartScreen tem: 14 valores diferentes
A coluna Firewall tem: 3 valores diferentes
A coluna UacLuaenable tem: 8 valores diferentes
A coluna Census_MDC2FormFactor tem: 12 valores diferentes
A coluna Census_DeviceFamily tem: 3 valores diferentes
A coluna Census_OEMNameIdentifier tem: 1665 valores diferentes
A coluna Census_OEMModelIdentifier tem: 45095 valores diferentes
A coluna Census_ProcessorCoreCount tem: 29 valores diferentes
A coluna Census_ProcessorManufacturerIdentifier tem: 5 valores diferentes
A coluna Census_ProcessorModelIdentifier tem: 2332 valores diferentes
A coluna Census_ProcessorClass tem: 4 valores diferentes
A coluna Census_PrimaryDiskTotalCapacity tem: 1194 valores diferentes
A coluna Census_PrimaryDiskTypeName tem: 5 valores diferentes
A coluna Census_SystemVolumeTotalCapacity tem: 158821 valores diferentes
A coluna Censu

### Resumo da análise às variáveis

#### Variáveis a serem removidas:
- Nulos:
    - **PuaMode**: 99.98% dos valores são nulos
    - **Census_ProcessorClass**: 99.59% dos valores são nulos e parece apresentar uma correlação muito elevada com o atributo "Census_ProcessorCoreCount"
- Muitos valores distintos:
    - **IeVerIdentifier**: Muitos valores distintos não agrupaveis (189)
    - **Census_OEMNameIdentifier**: Muitos valores distintos não agrupaveis (1664)
    - **Census_OEMModelIdentifier**: Muitos valores distintos não agrupaveis (45095)
    - **Census_ProcessorModelIdentifier**: Muitos valores distintos não agrupaveis (2331) e o atributo que identifica o fabricante parece ser mais interessante que o modelo do processador em si
    
#### Variáveis para alterar os valores NaN:
- (06.02% NaN) - **SMode**: substituir os missing por "0"
- (35.61% NaN) - **SmartScreen**: substituir os NaN pelo valor mais frequente ("RequireAdmin") ou "Off"?
- (01.03% NaN) - **Firewall**: substituir os NaN pelo valor mais frequente - "1" - (serviço ativo)
- (00.12% NaN) - **UacLuaenable**: substituir os NaN pelo valor mais frequente - "1" - (serviço ativo)
- (00.48% NaN) - **Census_ProcessorCoreCount**: substituir os NaN pela média/mediana dos restantes valores
- (00.48% NaN) - **Census_ProcessorManufacturerIdentifier**: substituir os NaN pelo valor mais frequente - "5"
- (00.62% NaN) - **Census_PrimaryDiskTotalCapacity**: substituir os NaN pela média/mediana dos restantes valores
- (00.15% NaN) - **Census_PrimaryDiskTypeName**: substituir os NaN pelo valor mais frequente - "HDD"
- (00.62% NaN) - **Census_SystemVolumeTotalCapacity**: substituir os NaN pela média/mediana dos restantes valores
- (00.94% NaN) - **Census_TotalPhysicalRAM**: substituir os NaN pela média/mediana dos restantes valores
- (00.01% NaN) - **Census_ChassisTypeName**: substituir os NaN pelo valor mais frequente - "Notebook"
- (00.54% NaN) - **Census_InternalPrimaryDiagonalDisplaySizeInInches**: substituir os NaN pela média dos restantes valores
- (00.54% NaN) - **Census_InternalPrimaryDisplayResolutionHorizontal**: substituir os NaN pela média/mediana dos restantes valores
- (00.54% NaN) - **Census_InternalPrimaryDisplayResolutionVertical**: substituir os NaN pela média/mediana dos restantes valores
- (0.001% NaN) - **Census_PowerPlatformRoleName**: substituir os valores missing pelo valor mais frequente - "Mobile"
- (71.05% NaN) - **Census_InternalBatteryType**: vale a pena substituir os NaN?
- (03.04% NaN) - **Census_InternalBatteryNumberOfCharges**: substituir os NaN pela média/mediana dos restantes valores

#### Variáveis para agrupar valores:
- **SmartScreen**: agrupar ["on", "On"] -> "On"  e ["off", "Off", "OFF"] -> "Off"
- **UacLuaenable**: agrupar todos os valores >= "1" em "1" (este atributo é um booleano)
- **Census_DeviceFamily**: agrupar ["Windows.Desktop", "Windows"] -> "Windows.Desktop"
- **Census_InternalBatteryType**: se ficar, agrupar [todos os que começam por "li"] -> "li"
- **Census_OSVersion**: agrupar pelo terceiro campo ["10.0.17134.228", "10.0.17134.165"] -> "10.0.17134"

#### Variáveis que aparentam possuir relações entre si:
- *Census_MDC2FormFactor*, *Census_ChassisTypeName*, 
- *Census_PrimaryDiskTotalCapacity*, *Census_SystemVolumeTotalCapacity*
- *Census_InternalPrimaryDiagonalDisplaySizeInInches*, *Census_InternalPrimaryDisplayResolutionHorizontal*, *Census_InternalPrimaryDisplayResolutionVertical*

In [37]:
def analisa_coluna(colName):
    coluna = ms[colName]
    
    percent_missing = coluna.isnull().sum() * 100 / nLinhas
    print("Percentagem NAs: " + str(percent_missing))
    
    valoresUnicos = coluna.unique()
    print("Valores diferentes: " + str(len(valoresUnicos)))
    
    print(coluna.value_counts(dropna=False))

### Drop das colunas 

In [53]:
cols = ['PuaMode', 'Census_ProcessorClass','IeVerIdentifier','Census_OEMModelIdentifier',
        'Census_OEMNameIdentifier', 'Census_InternalBatteryType']

cols = [c for c in cols if c in ms.columns]
ms = ms.drop(columns=cols)

ms.shape

(600000, 20)

### Tratar os NaN

In [54]:
ms['SMode'] = ms['SMode'].fillna(0)
ms['SmartScreen'] = ms['SmartScreen'].fillna('Off')
ms['Firewall'] = ms['Firewall'].fillna(1)
ms['UacLuaenable'] = ms['UacLuaenable'].fillna(1)
ms['Census_ProcessorCoreCount'] = ms['Census_ProcessorCoreCount'].fillna(ms['Census_ProcessorCoreCount'].mean())
ms['Census_ProcessorManufacturerIdentifier'] = ms['Census_ProcessorManufacturerIdentifier'].fillna(5)
ms['Census_PrimaryDiskTotalCapacity'] = ms['Census_PrimaryDiskTotalCapacity'].fillna(ms['Census_PrimaryDiskTotalCapacity'].mean())
ms['Census_PrimaryDiskTypeName'] = ms['Census_PrimaryDiskTypeName'].fillna('HDD')
ms['Census_SystemVolumeTotalCapacity'] = ms['Census_SystemVolumeTotalCapacity'].fillna(ms['Census_SystemVolumeTotalCapacity'].mean())
ms['Census_TotalPhysicalRAM'] = ms['Census_TotalPhysicalRAM'].fillna(ms['Census_TotalPhysicalRAM'].mean())
ms['Census_ChassisTypeName'] = ms['Census_ChassisTypeName'].fillna('Notebook')
ms['Census_InternalPrimaryDiagonalDisplaySizeInInches'] = ms['Census_InternalPrimaryDiagonalDisplaySizeInInches'].fillna(ms['Census_InternalPrimaryDiagonalDisplaySizeInInches'].mean())
ms['Census_InternalPrimaryDisplayResolutionHorizontal'] = ms['Census_InternalPrimaryDisplayResolutionHorizontal'].fillna(ms['Census_InternalPrimaryDisplayResolutionHorizontal'].mean())
ms['Census_InternalPrimaryDisplayResolutionVertical'] = ms['Census_InternalPrimaryDisplayResolutionVertical'].fillna(ms['Census_InternalPrimaryDisplayResolutionVertical'].mean())
ms['Census_PowerPlatformRoleName'] = ms['Census_PowerPlatformRoleName'].fillna('Mobile')
ms['Census_InternalBatteryNumberOfCharges'] = ms['Census_InternalBatteryNumberOfCharges'].fillna(ms['Census_InternalBatteryNumberOfCharges'].mean())


#### Verificar existência de NaNs:

In [58]:
soma=0
for col in ms.columns:
       soma = soma + ms[col].isnull().sum()

print("Nulls encontrados: " + str(soma))

Nulls encontrados: 0


### Agrupar valores

In [90]:
def trataString(string):
    words = string.split(".")
    palavra= '' + words[0] + '.' + words[1] + '.' + words[2]
    return palavra

In [91]:
ms['SmartScreen'] = ms['SmartScreen'].replace({"on": "On", "off": "Off", "OFF": "Off"})

ms.loc[ms['UacLuaenable'] > 1, 'UacLuaenable'] = 1

ms['Census_DeviceFamily'] = ms['Census_DeviceFamily'].replace({"Windows": "Windows.Desktop"})

ms['Census_OSVersion'] = ms['Census_OSVersion'].apply(lambda x : trataString(x))

### O que ainda falta fazer

In [ ]:
Verificar se vai ser para usar a variável Census_InternalBatteryType e tratá-la em caso afirmativo

Verificar relações entre diferentes variáveis

In [85]:
import scipy.stats as stats
from scipy.stats import chi2_contingency
import gc

def TestIndependence(colX,colY, alpha=0.05):
    aux = pd.crosstab(colY.astype(str),colX.astype(str)) 
    #print(aux)
    _, p, _, _ = stats.chi2_contingency(aux.values)
        
    #del aux
    gc.collect()
    return p

In [87]:
TestIndependence(ms['Census_MDC2FormFactor'],ms['Census_PowerPlatformRoleName'])

0.0

In [75]:
TestIndependence(ms['Census_PowerPlatformRoleName'],ms['Census_ChassisTypeName'])

0.0

In [89]:
TestIndependence(ms['Census_MDC2FormFactor'],ms['Census_ChassisTypeName'])

0.0